ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java

```diff
@@ -112,18 +137,84 @@ public void deleteSource(final String sourceName) {
   @Override
   public void putAll(MetaStore otherMetaStore) {
     this.topicMap.putAll(otherMetaStore.getAllKsqlTopics());
-    this.dataSourceMap.putAll(otherMetaStore.getAllStructuredDataSources());
+    this.dataSourceMap.putAll(otherMetaStore.getDataSourceMap());
+  }
+
+  @Override
+  public void addSourceNames(Set<String> sourceNames, String queryId) {
+    for (String sourceName: sourceNames) {
+      ReferentialIntegrityTableEntry referentialIntegrityTableEntry =
+          dataSourceMap.get(sourceName).getRight();
+      referentialIntegrityTableEntry.getSourceForQueries().add(queryId);
+    }
+  }
+
+  @Override
+  public void addSinkNames(Set<String> sinkNames, String queryId) {
+    for (String sinkName: sinkNames) {
+      ReferentialIntegrityTableEntry referentialIntegrityTableEntry =
+          dataSourceMap.get(sinkName).getRight();
+      referentialIntegrityTableEntry.getSinkForQueries().add(queryId);
+    }
+  }
+
+  @Override
+  public void removeQueryFromReferentialIntegrityTable(String queryId) {
+    for (Pair<StructuredDataSource, ReferentialIntegrityTableEntry>
+        structuredDataSourceReferentialIntegrityTableEntryPair: dataSourceMap.values()) {
+      structuredDataSourceReferentialIntegrityTableEntryPair.getRight()
```

Comment:
add a method to `ReferentialIntegrityTable`, i.e., `remove(queryId)`

Reply:
Done!